### PR TITLE
Rebase 1.6.0 beta3

### PR DESCRIPTION
**Pull request by ncdc**

### CODE REVIEW

File: pkg/cmd/server/kubernetes/master/master_config.go [hunk 1]
```diff
@@ -169,6 +170,7 @@ func BuildStorageFactory(masterConfig configapi.MasterConfig, server *kapiserver
 	if err != nil {
 		return nil, err
 	}
+	storageGroupsToEncodingVersion[autoscaling.GroupName] = kubecompatextensionsv1beta1.SchemeGroupVersion
```
Comment (sttts) [inline pkg/cmd/server/kubernetes/master/master_config.go]: Why this? Shouldn't we store them in the new format and enforce migration, and then in 1.7 finally remove extensionsv1beta1.HPA support?

File: pkg/cmd/server/etcd/etcd.go [hunk 2]
```diff
@@ -79,3 +80,56 @@ func TestEtcdClient(etcdClient etcdclient.Client) error {
...
+// connect to the etcd server and block until the server responds at least once, or return an
+// error if the server never responded.
+func TestEtcdClientV3(etcdClient *clientv3.Client) error {
+	for i := 0; ; i++ {
+		_, err := clientv3.NewKV(etcdClient).Get(context.Background(), "/", clientv3.WithLimit(1))
+		if err == nil || etcdutil.IsEtcdNotFound(err) {
+			break
+		}
+		if i > 100 {
+			return fmt.Errorf("could not reach etcd: %v", err)
```
Comment (sttts) [inline pkg/cmd/server/etcd/etcd.go:130]: This is 5sec. Enough?

File: pkg/cmd/server/kubernetes/master/master_config.go (hunk 1 above)
Comment (ncdc) [reply pkg/cmd/server/kubernetes/master/master_config.go]: Existing 3.5 clusters have existing HPAs encoded as extensions/v1beta1. Unless we make sure customers do a migration in 3.5.x, we have to keep them in extensions/v1beta1.

File: pkg/cmd/server/etcd/etcd.go (hunk 2 above)
Comment (ncdc) [reply pkg/cmd/server/etcd/etcd.go:130]: This is the equivalent of what's in master now.

File: test/integration/etcd_storage_path_test.go [hunk 3]
```diff
@@ -783,8 +877,26 @@ func TestEtcdStoragePath(t *testing.T) {
...
+						inGvkWhiteList = true
+						break
+					}
+				}
+				if inGvkWhiteList {
+					kindSeen.Insert(gvk.String())
+				} else {
+					kindSeen.Insert(kind)
+				}
+			*/
```
Comment (sttts) [inline test/integration/etcd_storage_path_test.go]: Can you put a comment above this block why it's commented out?

Comment (ncdc) [reply test/integration/etcd_storage_path_test.go]: I need to delete it

File: pkg/cmd/server/start/start_master.go [hunk 4]
```diff
@@ -479,15 +479,15 @@ func StartAPI(oc *origin.MasterConfig, kc *kubernetes.MasterConfig) error {
...
 	// verify we can connect to etcd with the provided config
-	if len(kc.Options.APIServerArguments) > 0 && len(kc.Options.APIServerArguments["storage-backend"]) > 0 && kc.Options.APIServerArguments["storage-backend"][0] == "etcd2" {
-		if _, err := etcd.GetAndTestEtcdClient(oc.Options.EtcdClientInfo); err != nil {
-			return err
-		}
-	} else {
-		if _, err := etcd.GetAndTestEtcdClientV3(oc.Options.EtcdClientInfo); err != nil {
-			return err
-		}
+	//if len(kc.Options.APIServerArguments) > 0 && len(kc.Options.APIServerArguments["storage-backend"]) > 0 && kc.Options.APIServerArguments["storage-backend"][0] == "etcd2" {
```
Comment (sttts) [inline pkg/cmd/server/start/start_master.go]: Should we keep this as a single commit? Maybe squash into the other commit with unread etcd3 changes?

File: pkg/cmd/server/kubernetes/master/master_config.go (hunk 1 above)
Comment (sttts) [reply pkg/cmd/server/kubernetes/master/master_config.go]: yes for reading/decoding. But won't this enforce that the whole autoscaling group is encoded with the extensions kind?

File: pkg/cmd/server/etcd/etcd.go (hunk 2 above)
Comment (sttts) [reply pkg/cmd/server/etcd/etcd.go:130]: ok, then it's probably fine.

**Comment by sttts** [comment]
Cherry-picked.

**Pull request closed without merging**